[gloo] transport open registration

jiayisuse opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#29164 [gloo] transport open registration**

- Created GlooDeviceFactory to hide device creation details
- Added transport option while on Python interface

The reason of making the factory class is to make it easier to extend gloo transport in the future

Differential Revision: [D18313537](https://our.internmc.facebook.com/intern/diff/D18313537)

jiayisuse committed:
[gloo] transport open registration

[ghstack-poisoned]

satgera commented on torch/distributed/distributed_c10d.py:
```diff
@@ -312,7 +312,8 @@ def init_process_group(backend,
                        store=None,
-                       group_name=''):
+                       group_name='',
+                       transport=''):
```
Why not reuse backend field like gloo-tls instead of adding a new param transport.

jiayisuse replied:
I see your point. Basically we want to keep the interface as it is. Will address

jiayisuse replied:
I see your point. Basically we want to keep the interface as it is. Available backends are hardcoded in torch/distributed/distributed_c10d.py as below. Adding GLOO-TLS may be tricky to isolate from OSS

  60     UNDEFINED = "undefined"
  61     GLOO = "gloo"
  62     NCCL = "nccl"
  63     MPI = "mpi"
  64     TCP = "tcp"

jiayisuse committed:
Update on "[gloo] transport open registration"


- Created GlooDeviceFactory to hide device creation details
- Added transport option while on Python interface

The reason of making the factory class is to make it easier to extend gloo transport in the future

Differential Revision: [D18313537](https://our.internmc.facebook.com/intern/diff/D18313537)

[ghstack-poisoned]

jiayisuse committed:
Update on "[gloo] transport open registration"


- Created GlooDeviceFactory to hide device creation details
- Added transport option while on Python interface

The reason of making the factory class is to make it easier to extend gloo transport in the future

Differential Revision: [D18313537](https://our.internmc.facebook.com/intern/diff/D18313537)

[ghstack-poisoned]

jiayisuse committed:
Update on "[gloo] transport open registration"


- Created GlooDeviceFactory to hide device creation details
- Added transport option while on Python interface

The reason of making the factory class is to make it easier to extend gloo transport in the future

Differential Revision: [D18313537](https://our.internmc.facebook.com/intern/diff/D18313537)

[ghstack-poisoned]

jiayisuse committed:
Update on "[gloo] transport open registration"


- Created GlooDeviceFactory to hide device creation details
- Added transport option while on Python interface

The reason of making the factory class is to make it easier to extend gloo transport in the future

Differential Revision: [D18313537](https://our.internmc.facebook.com/intern/diff/D18313537)

[ghstack-poisoned]

jiayisuse committed:
Update on "[gloo] transport open registration"


- Created GlooDeviceFactory to hide device creation details
- Added transport option while on Python interface

The reason of making the factory class is to make it easier to extend gloo transport in the future

Differential Revision: [D18313537](https://our.internmc.facebook.com/intern/diff/D18313537)

[ghstack-poisoned]

d4l3k commented on torch/lib/c10d/GlooDeviceFactory.cpp:
```diff
@@ -0,0 +1,103 @@
+std::shared_ptr<::gloo::transport::Device> GlooDeviceFactory::
+    makeDeviceForInterface(const std::string& interface) {
+#ifdef __linux__
+  return GlooDeviceRegistry()->Create("TCP", interface, "");
```
I'm a bit confused here so please let me know if I'm misunderstanding this.

What's the point of using the registry if you always access a specific implementation of it? Shouldn't this arg always be the same? From the fb specific diff it's just using a copy of this file w/ a different implementation of this method and different linking so it doesn't seem like you're using the registry overrides? If we're going to go that way might as well just remove all the registry code. 

cc @pietern

pietern replied:
+1

This should simply iterate over the available constructors and either 1) pick the first one, or 2) if an env var is set (e.g. `GLOO_TRANSPORT_TYPE`) pick that one.

Then there is no need to override this function and all you need to do is add a compilation unit where additional implementations are registered.

satgera replied:
Agree with @pietern and @d4l3k I was imagining a registry with key and priority. Pick the highest priority transport that's registered unless overwritten by an env which is the key to one of the registered transport. 
On the other hand since this its decided to be all an internal gloo implementation stuff with no interface change. I am fine doing it in an iteration over this.

jiayisuse replied:
@pietern Please correct me if I'm wrong. You are suggesting to check env in open source code?

jiayisuse replied:
@satgera To achieve this - pick up higher priority registration unless overwritten by env, I think we need to have two registries? One for platformRegistry with priorities, another one for transportRegistry without priorities.

jiayisuse replied:
To recap my thoughts a bit - 1) I'm going to hide any other specific logic from open source, and it's not necessarily to be a simple add-on or simple copy. For example, open source provides TCP and UV, but the stacked implementation provides TCP and TLS (no UV). So I didn't reuse the registry but implemented factory override. 2) env check is just for internal purpose to be introduced. Thus I isolated from open source by overriding in separate implementation. 3) Registry with priority is fine but will make TLS by default and enforce all customers do TLS. From my perspective, providing env configuration is preferred witch ends up with another question 4) Which transport are we going to default in the stacked implementation? We can discuss over that diff. 5) Given these thinking, at the first beginning I was also a bit confused why to use registry. Maybe we can get rid of this constrain to use registry.

Given all of the thoughts, I still think it makes more sense to have separate factory implementation with better isolation and flexibility.

satgera replied:
> @pietern Please correct me if I'm wrong. To just add additional registration in a separate source file, you are suggesting to check env in open source code?

Its fine the check env but the env should be to the effect GLOO_TRANSPORT_OVERRIDE = "tcp"/"uv"/"xx"

jiayisuse committed:
Update on "[gloo] transport open registration"


- Created GlooDeviceFactory to hide device creation details
- Added transport option while on Python interface

The reason of making the factory class is to make it easier to extend gloo transport in the future

Differential Revision: [D18313537](https://our.internmc.facebook.com/intern/diff/D18313537)

[ghstack-poisoned]

jiayisuse committed:
Update on "[gloo] transport open registration"


- Created GlooDeviceFactory to hide device creation details
- Added transport option while on Python interface

The reason of making the factory class is to make it easier to extend gloo transport in the future

Differential Revision: [D18313537](https://our.internmc.facebook.com/intern/diff/D18313537)

[ghstack-poisoned]

pietern commented on torch/lib/c10d/GlooDeviceFactory.cpp:
```diff
@@ -0,0 +1,117 @@
+  return ::gloo::transport::tcp::CreateDevice(attr);
+}
+
+C10_REGISTER_CREATOR(GlooDeviceRegistry, LINUX, makeTCPDevice);
```
Why register it twice? If there is a good reason please document it in a comment here.

pietern commented on torch/lib/c10d/GlooDeviceFactory.cpp:
```diff
@@ -0,0 +1,117 @@
+  return ::gloo::transport::uv::CreateDevice(attr);
+}
+
+C10_REGISTER_CREATOR(GlooDeviceRegistry, APPLE, makeUVDevice);
```
Same.

pietern commented on torch/lib/c10d/GlooDeviceFactory.cpp:
```diff
@@ -0,0 +1,117 @@
+C10_REGISTER_CREATOR(GlooDeviceRegistry, UV, makeUVDevice);
+#endif
+
+static const char* glooDevice = getenv("GLOO_DEVICE");
```
This should be `GLOO_DEVICE_TRANSPORT` or `GLOO_TRANSPORT`.

pietern reviewed (commented):
A couple minor comments. Also, can you use lower case identifiers for keys.

jiayisuse committed:
Update on "[gloo] transport open registration"


- Created GlooDeviceFactory to hide device creation details
- Added transport option while on Python interface

The reason of making the factory class is to make it easier to extend gloo transport in the future

Differential Revision: [D18313537](https://our.internmc.facebook.com/intern/diff/D18313537)

[ghstack-poisoned]

jiayisuse committed:
Update on "[gloo] transport open registration"


- Created GlooDeviceFactory to hide device creation details
- Added transport option while on Python interface

The reason of making the factory class is to make it easier to extend gloo transport in the future

Differential Revision: [D18313537](https://our.internmc.facebook.com/intern/diff/D18313537)

[ghstack-poisoned]
